server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -276,10 +280,12 @@ private static class WeightFunction {
         }
 
         float weight(Balancer balancer, ModelNode node, String index) {
-            final float weightShard = node.numShards() - balancer.avgShardsPerNode();
+            final var tier = balancer.indexTier(index);
+            final var tierStats = balancer.getStatsForTier(tier);
+            final float weightShard = node.numShards() - tierStats.avgShardSize();
             final float weightIndex = node.numShards(index) - balancer.avgShardsPerNode(index);
```

Comment:
`balancer.avgShardsPerNode(index)` -> `((float) balancer.metadata.index(index).getTotalNumberOfShards() / tierStats.numberOfNodes)` so that it is also tier relative